homeassistant/components/apple_tv/media_player.py

```diff
@@ -282,22 +282,20 @@ async def async_play_media(self, media_type, media_id, **kwargs):
         # RAOP. Otherwise try to play it with regular AirPlay.
         if media_type == MEDIA_TYPE_APP:
             await self.atv.apps.launch_app(media_id)
+            return
 
         if media_source.is_media_source_id(media_id):
             play_item = await media_source.async_resolve_media(
                 self.hass, media_id, self.entity_id
             )
-            media_id = play_item.url
+            media_id = async_process_play_media_url(self.hass, play_item.url)
```

Comment:
I don't think this is right? How would this handle signing for non-media sources now?

The reported issue is raising `ValueError("URL is relative, but does not start with a /")`, which is odd it raised in the first places. Where do these non-root relatives come from?

Reply:
The issue I noticed was that if `media_id` is passed a local filename, e.g. /media/myfile.mp3, it would always get translated into a media source URL (https). That's why I feel it's wrong in the general case to call `async_process_play_media_url` , so I moved it.

Reply:
I'm not following your logic on that one? It will translate it to `https` in case the relative URL is not explicitly allowed AND the internal URL isn't available (or the internal URL itself it set to `https`).

Isn't the solution to allow for relative URLs instead?

`async_process_play_media_url(self.hass, play_item.url, allow_relative_url: true)`?

Reply:
Right, I didn't know about `allow_relative_url` but that makes more sense. A problem however is that it signs the URL:

```
2022-06-28 11:20:32 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140147523320272] Error handling message: Unknown error (unknown_error)
Traceback (most recent call last):
  File "/ha/homeassistant/components/websocket_api/decorators.py", line 27, in _handle_async_response
    await func(hass, connection, msg)
  File "/ha/homeassistant/components/websocket_api/commands.py", line 636, in handle_execute_script
    await script_obj.async_run(msg.get("variables"), context=context)
  File "/ha/homeassistant/helpers/script.py", line 1513, in async_run
    await asyncio.shield(run.async_run())
  File "/ha/homeassistant/helpers/script.py", line 405, in async_run
    await self._async_step(log_exceptions=False)
  File "/ha/homeassistant/helpers/script.py", line 449, in _async_step
    self._handle_exception(
  File "/ha/homeassistant/helpers/script.py", line 472, in _handle_exception
    raise exception
  File "/ha/homeassistant/helpers/script.py", line 447, in _async_step
    await getattr(self, handler)()
  File "/ha/homeassistant/helpers/script.py", line 680, in _async_call_service_step
    await service_task
  File "/ha/homeassistant/core.py", line 1713, in async_call
    task.result()
  File "/ha/homeassistant/core.py", line 1750, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/ha/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await service.entity_service_call(
  File "/ha/homeassistant/helpers/service.py", line 680, in entity_service_call
    future.result()  # pop exception if have
  File "/ha/homeassistant/helpers/entity.py", line 900, in async_request_call
    await coro
  File "/ha/homeassistant/helpers/service.py", line 717, in _handle_entity_call
    await result
  File "/ha/homeassistant/components/apple_tv/media_player.py", line 300, in async_play_media
    await self.atv.stream.stream_file(media_id)
  File "/ha/venv/lib/python3.10/site-packages/pyatv/core/facade.py", line 382, in stream_file
    await self.relay("stream_file")(file, **kwargs)
  File "/ha/venv/lib/python3.10/site-packages/pyatv/protocols/raop/__init__.py", line 371, in stream_file
    audio_file = await open_source(
  File "/ha/venv/lib/python3.10/site-packages/pyatv/protocols/raop/audio_source.py", line 484, in open_source
    return await FileSource.open(source, sample_rate, channels, sample_size)
  File "/ha/venv/lib/python3.10/site-packages/pyatv/protocols/raop/audio_source.py", line 428, in open
    src = await loop.run_in_executor(
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/ha/venv/lib/python3.10/site-packages/miniaudio.py", line 929, in decode_file
    filenamebytes = _get_filename_bytes(filename)
  File "/ha/venv/lib/python3.10/site-packages/miniaudio.py", line 807, in _get_filename_bytes
    raise FileNotFoundError(filename)
FileNotFoundError: /ha/sample.wav?authSig=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiIwOGMzMTBhMTdhMDc0Y2JmODlhOTg4NGY3NTkzOTYyNCIsInBhdGgiOiIvaGEvc2FtcGxlLndhdiIsInBhcmFtcyI6e30sImlhdCI6MTY1NjQxNTIzMSwiZXhwIjoxNjU2NTAxNjMxfQ.49LjId3xmWoY5stK2uoA-epyGT0wQb5MRCoCLgVV6ZI
```

Reply:
Right, so it's not sent over as an URL to the device (which fetches and plays it), but instead, we try to open the file and stream its contents?

If so, why? as in, looking at the above, it can handle URLs?

Reply:
Yeah, pyatv opens the file locally and streams it. Audio streaming with AirPlay is purely local. A receiver cannot pull stream a stream from a URL itself, a sender must stream audio packets to it. So streaming from a URL basically means that Home Assiatant serves a file that pyatv downloads (streams), decodes and sends to the receiver. That loop is very excessive when pyatv can just open the file locally instead. Also, streaming via HTTP is even more experimental than regular streaming from a file. It has a smaller buffer and less poorer error handling, so in reality I would prefer steaming over HTTP altogether.

Reply:
> So streaming from a URL basically means that Home Assiatant serves a file that pyatv downloads (streams), decodes and sends to the receiver. That loop is very excessive when pyatv can just open the file locally instead.

So why do we generate URLs out of media sources in that case (which you moved and kept)?

Reply:
> So why do we generate URLs out of media sources in that case (which you moved and kept)?

From what I have gathered, a media id can resolve to a local file (which is served via the web server, so it generates a URL) but it could also resolve to a URL by itself in case of it being a remote resource (like a URL to an Internet radio station for instance). So a media id must be resolvable for these cases. A local file would just be a local file and should not be resolved in any way. If a media id for a resource in the media library can be resolved to local file, then I would prefer to get the file path instead to avoid looping data over HTTP but still resolve in other cases and fallback to URL-based when nothing else is possible. Locally I have reverted the change and made it to resolve relative, but not pushed it (until I know it's actually supposed to be).

Reply:
Media sources always resolve to URLs. If you know of a media source that doesn't, that's a bug and we should fix that.

If a media source resolves to a relative URL, it needs to be prefixed with the HA host. That's what `async_process_play_media_url` is for. 

If ATV supports local file paths, moving `async_process_play_media_url` into the browse media if-statement is correct. That way it will only apply to the output of resolved media sources and not accidentally to local file paths.

We suggest putting `async_process_play_media_url` outside of the if-statement to allow passing any relative HA URL to play_media and it works. In this case this move inside the if-statement is fine as we cannot distinguish between relative HTTP urls and file paths.